File: apps/dukcapil-pbd-be/internal/repository/data_wilayah.go
package repository

import (
	"context"
	"errors"
	"fmt"
	"strings"
	"time"

	"dukcapil-pbd-be/internal/model"

	"gorm.io/gorm"
)

type DataWilayahRepository struct {
	db *gorm.DB
}

func NewDataWilayahRepository(db *gorm.DB) *DataWilayahRepository {
	return &DataWilayahRepository{db: db}
}

func (r *DataWilayahRepository) List(ctx context.Context, tahunAnggaran string) (model.DataWilayahResponse, error) {
	db, err := r.session(ctx)
	if err != nil {
		return model.DataWilayahResponse{}, err
	}

	tahunAnggaran = strings.TrimSpace(tahunAnggaran)

	var records []model.DataWilayahEntity
	if err := db.Where("tahun_anggaran = ?", tahunAnggaran).Order("sort_order ASC").Find(&records).Error; err != nil {
		return model.DataWilayahResponse{}, fmt.Errorf("list data wilayah: %w", err)
	}

	regions := make([]model.RegionData, 0, len(records))
	latestUpdatedAt := time.Time{}
	for _, record := range records {
		regions = append(regions, record.ToRegionData())
		if record.UpdatedAt.After(latestUpdatedAt) {
			latestUpdatedAt = record.UpdatedAt
		}
	}
	if err := r.attachBumKampungSummary(ctx, db, tahunAnggaran, regions); err != nil {
		return model.DataWilayahResponse{}, err
	}

	var updatedAt *time.Time
	if !latestUpdatedAt.IsZero() {
		updatedAt = &latestUpdatedAt
	}

	return model.DataWilayahResponse{
		TahunAnggaran: tahunAnggaran,
		Regions:       regions,
		UpdatedAt:     updatedAt,
	}, nil
}

func (r *DataWilayahRepository) UpdateDukcapil(
	ctx context.Context,
	tahunAnggaran string,
	id string,
	payload model.DataWilayahDukcapilPayload,
) (model.RegionData, bool, error) {
	db, err := r.session(ctx)
	if err != nil {
		return model.RegionData{}, false, err
	}

	tahunAnggaran = strings.TrimSpace(tahunAnggaran)
	id = strings.TrimSpace(id)

	jumlahJiwa := int64(payload.Oap.JumlahOap) + int64(payload.Oap.JumlahNonOap)
	if jumlahJiwa < 0 || jumlahJiwa > 2147483647 {
		return model.RegionData{}, false, fmt.Errorf("jumlah jiwa is out of range")
	}

	var record model.DataWilayahEntity
	err = db.Transaction(func(tx *gorm.DB) error {
		result := tx.Model(&model.DataWilayahEntity{}).
			Where("tahun_anggaran = ? AND id = ?", tahunAnggaran, id).
			Updates(map[string]any{
				"registration_penerbitan_kk":     payload.Registration.PenerbitanKk,
				"registration_perubahan_kk":      payload.Registration.PerubahanKk,
				"registration_kia":               payload.Registration.Kia,
				"registration_nik_wni":           payload.Registration.NikWni,
				"registration_perekaman_ktp_el":  payload.Registration.PerekamanKtpEl,
				"registration_pencetakan_ktp_el": payload.Registration.PencetakanKtpEl,
				"oap_luas_wilayah":               payload.Oap.LuasWilayah,
				"oap_jumlah_oap":                 payload.Oap.JumlahOap,
				"oap_jumlah_non_oap":             payload.Oap.JumlahNonOap,
				"oap_jumlah_jiwa":                int(jumlahJiwa),
				"civil_akta_kelahiran":           payload.Civil.AktaKelahiran,
				"civil_akta_kematian":            payload.Civil.AktaKematian,
				"civil_akta_perkawinan":          payload.Civil.AktaPerkawinan,
				"civil_akta_perceraian":          payload.Civil.AktaPerceraian,
				"updated_at":                     gorm.Expr("NOW()"),
			})
		if result.Error != nil {
			return result.Error
		}
		if result.RowsAffected == 0 {
			return gorm.ErrRecordNotFound
		}

		return tx.First(&record, "tahun_anggaran = ? AND id = ?", tahunAnggaran, id).Error
	})
	if errors.Is(err, gorm.ErrRecordNotFound) {
		return model.RegionData{}, false, nil
	}
	if err != nil {
		return model.RegionData{}, false, fmt.Errorf("update data wilayah Dukcapil: %w", err)
	}

	return record.ToRegionData(), true, nil
}

func (r *DataWilayahRepository) GetWebsiteSettings(ctx context.Context) (model.DataWilayahWebsiteSettingsResponse, error) {
	db, err := r.session(ctx)
	if err != nil {
		return model.DataWilayahWebsiteSettingsResponse{}, err
	}

	availableYears, err := r.listAvailableYears(db)
	if err != nil {
		return model.DataWilayahWebsiteSettingsResponse{}, err
	}
	if len(availableYears) == 0 {
		return model.DataWilayahWebsiteSettingsResponse{
			FeaturedTahunAnggaran:  "",
			PublishedTahunAnggaran: []string{},
			AvailableTahunAnggaran: []string{},
		}, nil
	}

	entity, changed, err := r.loadOrCreateWebsiteSettings(db, availableYears)
	if err != nil {
		return model.DataWilayahWebsiteSettingsResponse{}, err
	}
	if changed {
		if err := db.Save(&entity).Error; err != nil {
			return model.DataWilayahWebsiteSettingsResponse{}, fmt.Errorf("persist data wilayah settings: %w", err)
		}
	}

	settings := entity.ToWebsiteSettings()
	return model.DataWilayahWebsiteSettingsResponse{
		FeaturedTahunAnggaran:  settings.FeaturedTahunAnggaran,
		PublishedTahunAnggaran: settings.PublishedTahunAnggaran,
		AvailableTahunAnggaran: availableYears,
	}, nil
}

func (r *DataWilayahRepository) UpdateWebsiteSettings(
	ctx context.Context,
	payload model.DataWilayahWebsiteSettingsPayload,
) (model.DataWilayahWebsiteSettingsResponse, error) {
	db, err := r.session(ctx)
	if err != nil {
		return model.DataWilayahWebsiteSettingsResponse{}, err
	}

	availableYears, err := r.listAvailableYears(db)
	if err != nil {
		return model.DataWilayahWebsiteSettingsResponse{}, err
	}
	if len(availableYears) == 0 {
		return model.DataWilayahWebsiteSettingsResponse{}, fmt.Errorf("data wilayah belum tersedia")
	}

	publishedYears := normalizeYears(payload.PublishedTahunAnggaran, availableYears)
	if len(publishedYears) == 0 {
		return model.DataWilayahWebsiteSettingsResponse{}, fmt.Errorf("minimal satu tahun release wajib dipilih")
	}

	featuredYear := strings.TrimSpace(payload.FeaturedTahunAnggaran)
	if !containsYear(publishedYears, featuredYear) {
		return model.DataWilayahWebsiteSettingsResponse{}, fmt.Errorf("tahun unggulan wajib termasuk dalam tahun release")
	}

	entity, _, err := r.loadOrCreateWebsiteSettings(db, availableYears)
	if err != nil {
		return model.DataWilayahWebsiteSettingsResponse{}, err
	}

	entity.FeaturedTahunAnggaran = featuredYear
	entity.PublishedTahunAnggaran = publishedYears

	if err := db.Save(&entity).Error; err != nil {
		return model.DataWilayahWebsiteSettingsResponse{}, fmt.Errorf("update data wilayah settings: %w", err)
	}

	return model.DataWilayahWebsiteSettingsResponse{
		FeaturedTahunAnggaran:  entity.FeaturedTahunAnggaran,
		PublishedTahunAnggaran: append([]string(nil), entity.PublishedTahunAnggaran...),
		AvailableTahunAnggaran: availableYears,
	}, nil
}

func (r *DataWilayahRepository) listAvailableYears(db *gorm.DB) ([]string, error) {
	currentYear := currentDataWilayahReleaseYear()
	if err := r.ensureCurrentReleaseYearData(db, currentYear); err != nil {
		return nil, err
	}

	var years []string
	if err := db.Model(&model.DataWilayahEntity{}).
		Distinct("tahun_anggaran").
		Where("tahun_anggaran <= ?", currentYear).
		Order("tahun_anggaran DESC").
		Pluck("tahun_anggaran", &years).Error; err != nil {
		return nil, fmt.Errorf("list data wilayah years: %w", err)
	}

	return years, nil
}

func (r *DataWilayahRepository) ensureCurrentReleaseYearData(db *gorm.DB, currentYear string) error {
	var count int64
	if err := db.Model(&model.DataWilayahEntity{}).
		Where("tahun_anggaran = ?", currentYear).
		Count(&count).Error; err != nil {
		return fmt.Errorf("count current data wilayah year: %w", err)
	}
	if count > 0 {
		return nil
	}

	var sourceYear string
	if err := db.Model(&model.DataWilayahEntity{}).
		Select("tahun_anggaran").
		Where("tahun_anggaran < ?", currentYear).
		Order("tahun_anggaran DESC").
		Limit(1).
		Scan(&sourceYear).Error; err != nil {
		return fmt.Errorf("find current release source data wilayah: %w", err)
	}
	if sourceYear == "" {
		return nil
	}

	if err := db.Exec(`
		INSERT INTO data_wilayah (
			tahun_anggaran, id, sort_order, name, short_name, region_type, map_label,
			idm_sangat_tertinggal, idm_tertinggal, idm_berkembang, idm_maju, idm_mandiri,
			bumdes_jumlah, bumdes_aktif, bumdes_tidak_aktif, bumdes_bersama,
			registration_penerbitan_kk, registration_perubahan_kk, registration_kia, registration_nik_wni,
			registration_perekaman_ktp_el, registration_pencetakan_ktp_el,
			oap_luas_wilayah, oap_jumlah_oap, oap_jumlah_non_oap, oap_jumlah_jiwa,
			civil_akta_kelahiran, civil_akta_kematian, civil_akta_perkawinan, civil_akta_perceraian
		)
		SELECT
			?, id, sort_order, name, short_name, region_type, map_label,
			idm_sangat_tertinggal, idm_tertinggal, idm_berkembang, idm_maju, idm_mandiri,
			bumdes_jumlah, bumdes_aktif, bumdes_tidak_aktif, bumdes_bersama,
			registration_penerbitan_kk, registration_perubahan_kk, registration_kia, registration_nik_wni,
			registration_perekaman_ktp_el, registration_pencetakan_ktp_el,
			oap_luas_wilayah, oap_jumlah_oap, oap_jumlah_non_oap, oap_jumlah_jiwa,
			civil_akta_kelahiran, civil_akta_kematian, civil_akta_perkawinan, civil_akta_perceraian
		FROM data_wilayah
		WHERE tahun_anggaran = ?
		ON CONFLICT (tahun_anggaran, id) DO NOTHING
	`, currentYear, sourceYear).Error; err != nil {
		return fmt.Errorf("seed current data wilayah year: %w", err)
	}

	return nil
}

func currentDataWilayahReleaseYear() string {
	wit := time.FixedZone("WIT", 9*60*60)
	return fmt.Sprintf("%d", time.Now().In(wit).Year())
}

type bumKampungRegionSummary struct {
	KabupatenKota string
	Jumlah        int
	Aktif         int
	TidakAktif    int
	Bersama       int
}

func (r *DataWilayahRepository) attachBumKampungSummary(
	ctx context.Context,
	db *gorm.DB,
	tahunAnggaran string,
	regions []model.RegionData,
) error {
	if len(regions) == 0 {
		return nil
	}

	var summaries []bumKampungRegionSummary
	if err := db.WithContext(ctx).
		Table("bum_kampung").
		Select(`
			kabupaten_kota,
			COUNT(*)::INTEGER AS jumlah,
			COUNT(*) FILTER (
				WHERE status IN ('Dokumen Badan Hukum Terverifikasi', 'Nama Terverifikasi')
			)::INTEGER AS aktif,
			COUNT(*) FILTER (
				WHERE status NOT IN ('Dokumen Badan Hukum Terverifikasi', 'Nama Terverifikasi')
			)::INTEGER AS tidak_aktif,
			COUNT(*) FILTER (WHERE kategori = 'BUMKam bersama')::INTEGER AS bersama
		`).
		Where("tahun_anggaran = ?", strings.TrimSpace(tahunAnggaran)).
		Group("kabupaten_kota").
		Scan(&summaries).Error; err != nil {
		return fmt.Errorf("summarize bum kampung for data wilayah: %w", err)
	}

	applyBumKampungSummary(regions, summaries)
	return nil
}

func applyBumKampungSummary(
	regions []model.RegionData,
	summaries []bumKampungRegionSummary,
) {
	summaryByKabupaten := make(map[string]bumKampungRegionSummary, len(summaries))
	for _, summary := range summaries {
		summaryByKabupaten[normalizeRegionName(summary.KabupatenKota)] = summary
	}

	for index := range regions {
		regionKey := normalizeRegionName(regions[index].Name)
		summary, found := summaryByKabupaten[regionKey]
		if !found {
			regionKey = normalizeRegionName(regions[index].ShortName)
			summary, found = summaryByKabupaten[regionKey]
		}
		if !found {
			regions[index].Bumdes = model.BumdesData{}
			continue
		}

		regions[index].Bumdes = model.BumdesData{
			Jumlah:     summary.Jumlah,
			Aktif:      summary.Aktif,
			TidakAktif: summary.TidakAktif,
			Bersama:    summary.Bersama,
		}
	}
}

func normalizeRegionName(value string) string {
	normalized := strings.ToLower(strings.TrimSpace(value))
	normalized = strings.TrimPrefix(normalized, "kabupaten ")
	normalized = strings.TrimPrefix(normalized, "kota ")
	return strings.Join(strings.Fields(normalized), " ")
}

func (r *DataWilayahRepository) loadOrCreateWebsiteSettings(
	db *gorm.DB,
	availableYears []string,
) (model.DataWilayahPublicSettingsEntity, bool, error) {
	var entity model.DataWilayahPublicSettingsEntity
	err := db.First(&entity, "id = ?", 1).Error
	if errors.Is(err, gorm.ErrRecordNotFound) {
		entity = model.DataWilayahPublicSettingsEntity{
			ID:                     1,
			FeaturedTahunAnggaran:  availableYears[0],
			PublishedTahunAnggaran: availableYears,
		}
		if err := db.Create(&entity).Error; err != nil {
			return model.DataWilayahPublicSettingsEntity{}, false, fmt.Errorf("create data wilayah settings: %w", err)
		}
		return entity, false, nil
	}
	if err != nil {
		return model.DataWilayahPublicSettingsEntity{}, false, fmt.Errorf("load data wilayah settings: %w", err)
	}

	normalizedPublished := normalizeYears(entity.PublishedTahunAnggaran, availableYears)
	changed := false
	if len(normalizedPublished) == 0 {
		normalizedPublished = []string{availableYears[0]}
		changed = true
	}
	if !containsYear(availableYears, entity.FeaturedTahunAnggaran) ||
		!containsYear(normalizedPublished, entity.FeaturedTahunAnggaran) {
		entity.FeaturedTahunAnggaran = normalizedPublished[0]
		changed = true
	}
	if strings.Join(entity.PublishedTahunAnggaran, ",") != strings.Join(normalizedPublished, ",") {
		entity.PublishedTahunAnggaran = normalizedPublished
		changed = true
	}

	return entity, changed, nil
}

func normalizeYears(years []string, allowedYears []string) []string {
	result := make([]string, 0, len(years))
	seen := make(map[string]struct{}, len(years))
	for _, year := range allowedYears {
		for _, candidate := range years {
			normalized := strings.TrimSpace(candidate)
			if normalized != year {
				continue
			}
			if _, exists := seen[normalized]; exists {
				break
			}
			result = append(result, normalized)
			seen[normalized] = struct{}{}
			break
		}
	}

	return result
}

func containsYear(years []string, target string) bool {
	target = strings.TrimSpace(target)
	for _, year := range years {
		if strings.TrimSpace(year) == target {
			return true
		}
	}

	return false
}

func (r *DataWilayahRepository) session(ctx context.Context) (*gorm.DB, error) {
	if r == nil || r.db == nil {
		return nil, fmt.Errorf("database connection is required")
	}
	return r.db.WithContext(ctx), nil
}

File: apps/dukcapil-pbd-fe/lib/data-wilayah.ts
import type { RegionData } from "@/types/data-wilayah";

export type RegionTab = "registration" | "civil" | "oap" | "idm" | "bumdes";

export const defaultRegionData: RegionData[] = [
  {
    id: "kabupaten-sorong",
    name: "Kabupaten Sorong",
    shortName: "Sorong",
    type: "Kabupaten",
    mapLabel: "Kab. Sorong",
    idm: { sangatTertinggal: 60, tertinggal: 80, berkembang: 66, maju: 3, mandiri: 0 },
    bumdes: { jumlah: 0, aktif: 0, tidakAktif: 0, bersama: 0 },
    registration: { penerbitanKk: 4311, perubahanKk: 8945, kia: 3861, nikWni: 2756, perekamanKtpEl: 2511, pencetakanKtpEl: 21637 },
    oap: { luasWilayah: 6544.23, jumlahOap: 54379, jumlahNonOap: 76322, jumlahJiwa: 130701 },
    civil: { aktaKelahiran: 4213, aktaKematian: 947, aktaPerkawinan: 390, aktaPerceraian: 21 },
  },
  {
    id: "kota-sorong",
    name: "Kota Sorong",
    shortName: "Kota Sorong",
    type: "Kota",
    mapLabel: "Kota Sorong",
    idm: { sangatTertinggal: 0, tertinggal: 0, berkembang: 0, maju: 0, mandiri: 0 },
    bumdes: { jumlah: 0, aktif: 0, tidakAktif: 0, bersama: 0 },
    registration: { penerbitanKk: 9376, perubahanKk: 14612, kia: 1490, nikWni: 4627, perekamanKtpEl: 4418, pencetakanKtpEl: 27136 },
    oap: { luasWilayah: 656.64, jumlahOap: 77487, jumlahNonOap: 209765, jumlahJiwa: 287252 },
    civil: { aktaKelahiran: 7208, aktaKematian: 1941, aktaPerkawinan: 1118, aktaPerceraian: 47 },
  },
  {
    id: "raja-ampat",
    name: "Kabupaten Raja Ampat",
    shortName: "Raja Ampat",
    type: "Kabupaten",
    mapLabel: "Raja Ampat",
    idm: { sangatTertinggal: 16, tertinggal: 33, berkembang: 75, maju: 6, mandiri: 0 },
    bumdes: { jumlah: 0, aktif: 0, tidakAktif: 0, bersama: 0 },
    registration: { penerbitanKk: 2688, perubahanKk: 4543, kia: 1997, nikWni: 1713, perekamanKtpEl: 1466, pencetakanKtpEl: 10624 },
    oap: { luasWilayah: 8034.44, jumlahOap: 53035, jumlahNonOap: 20713, jumlahJiwa: 73748 },
    civil: { aktaKelahiran: 3998, aktaKematian: 545, aktaPerkawinan: 741, aktaPerceraian: 2 },
  },
  {
    id: "sorong-selatan",
    name: "Kabupaten Sorong Selatan",
    shortName: "Sorong Selatan",
    type: "Kabupaten",
    mapLabel: "Sorong Selatan",
    idm: { sangatTertinggal: 28, tertinggal: 40, berkembang: 73, maju: 4, mandiri: 0 },
    bumdes: { jumlah: 0, aktif: 0, tidakAktif: 0, bersama: 0 },
    registration: { penerbitanKk: 1342, perubahanKk: 2568, kia: 680, nikWni: 1570, perekamanKtpEl: 880, pencetakanKtpEl: 6031 },
    oap: { luasWilayah: 6594.31, jumlahOap: 46829, jumlahNonOap: 10684, jumlahJiwa: 57513 },
    civil: { aktaKelahiran: 2571, aktaKematian: 323, aktaPerkawinan: 359, aktaPerceraian: 5 },
  },
  {
    id: "maybrat",
    name: "Kabupaten Maybrat",
    shortName: "Maybrat",
    type: "Kabupaten",
    mapLabel: "Maybrat",
    idm: { sangatTertinggal: 107, tertinggal: 128, berkembang: 59, maju: 1, mandiri: 0 },
    bumdes: { jumlah: 0, aktif: 0, tidakAktif: 0, bersama: 0 },
    registration: { penerbitanKk: 1230, perubahanKk: 3222, kia: 190, nikWni: 696, perekamanKtpEl: 505, pencetakanKtpEl: 5220 },
    oap: { luasWilayah: 5461.69, jumlahOap: 43178, jumlahNonOap: 3626, jumlahJiwa: 46804 },
    civil: { aktaKelahiran: 1775, aktaKematian: 300, aktaPerkawinan: 203, aktaPerceraian: 4 },
  },
  {
    id: "tambrauw",
    name: "Kabupaten Tambrauw",
    shortName: "Tambrauw",
    type: "Kabupaten",
    mapLabel: "Tambrauw",
    idm: { sangatTertinggal: 202, tertinggal: 64, berkembang: 19, maju: 0, mandiri: 0 },
    bumdes: { jumlah: 0, aktif: 0, tidakAktif: 0, bersama: 0 },
    registration: { penerbitanKk: 631, perubahanKk: 1253, kia: 1025, nikWni: 596, perekamanKtpEl: 330, pencetakanKtpEl: 2861 },
    oap: { luasWilayah: 11529.18, jumlahOap: 21302, jumlahNonOap: 10086, jumlahJiwa: 31388 },
    civil: { aktaKelahiran: 830, aktaKematian: 120, aktaPerkawinan: 101, aktaPerceraian: 1 },
  },
];

export const tabLabels: Record<RegionTab, string> = {
  registration: "Pendaftaran Penduduk",
  civil: "Pencatatan Sipil",
  oap: "Data OAP",
  idm: "IDM Desa",
  bumdes: "Data BUMDes",
};

export const formatNumber = (value: number) =>
  new Intl.NumberFormat("id-ID").format(value);

export const formatArea = (value: number) =>
  `${new Intl.NumberFormat("id-ID", {
    maximumFractionDigits: 2,
    minimumFractionDigits: 2,
  }).format(value)} km²`;

export const getTotalIdmVillages = (idm: RegionData["idm"]) =>
  idm.sangatTertinggal + idm.tertinggal + idm.berkembang + idm.maju + idm.mandiri;

export const getTotalBumdes = (bumdes: RegionData["bumdes"]) =>
  bumdes.jumlah;

export const getDominantIdmStatus = (idm: RegionData["idm"]) => {
  const entries: Array<[string, number]> = [
    ["Sangat Tertinggal", idm.sangatTertinggal],
    ["Tertinggal", idm.tertinggal],
    ["Berkembang", idm.berkembang],
    ["Maju", idm.maju],
    ["Mandiri", idm.mandiri],
  ];
  const [status, total] = entries.reduce((highest, current) =>
    current[1] > highest[1] ? current : highest,
  );
  return total > 0 ? status : "Tidak tersedia";
};

export const getProvinceTotals = (regions: RegionData[]) => ({
  totalJiwa: regions.reduce((total, region) => total + region.oap.jumlahJiwa, 0),
  totalOap: regions.reduce((total, region) => total + region.oap.jumlahOap, 0),
  totalNonOap: regions.reduce((total, region) => total + region.oap.jumlahNonOap, 0),
  totalKtpEl: regions.reduce((total, region) => total + region.registration.pencetakanKtpEl, 0),
  totalDesaIdm: regions.reduce((total, region) => total + getTotalIdmVillages(region.idm), 0),
  totalBumdes: regions.reduce((total, region) => total + getTotalBumdes(region.bumdes), 0),
});
